File: Drawing/Gfx.cs
using System.Drawing.Drawing2D;

namespace AquariumTaskbar.Drawing;

internal static class Gfx
{
    public static void FillRound(Graphics graphics, Rectangle bounds, int radius, Color top, Color bottom)
    {
        if (bounds.Width <= 0 || bounds.Height <= 0)
        {
            return;
        }

        using var path = RoundedRect(bounds, radius);
        using var fill = new LinearGradientBrush(bounds, top, bottom, LinearGradientMode.Vertical);
        graphics.FillPath(fill, path);
    }

    public static void GlassCard(Graphics graphics, Rectangle bounds, int radius)
    {
        using (var shadow = RoundedRect(new Rectangle(bounds.X, bounds.Y + 3, bounds.Width, bounds.Height), radius))
        using (var shadowFill = new SolidBrush(Color.FromArgb(35, 12, 28, 45)))
        {
            graphics.FillPath(shadowFill, shadow);
        }

        FillRound(graphics, bounds, radius, Palette.CardTop, Palette.CardBottom);

        using var path = RoundedRect(bounds, radius);
        using var edge = new Pen(Color.FromArgb(195, 90, 150, 210), 1.4f);
        graphics.DrawPath(edge, path);

        using var innerPath = RoundedRect(Rectangle.Inflate(bounds, -2, -2), Math.Max(2, radius - 2));
        using var innerEdge = new Pen(Color.FromArgb(60, 70, 130, 170), 0.8f);
        graphics.DrawPath(innerEdge, innerPath);

        var state = graphics.Save();
        graphics.SetClip(path);
        var shine = new Rectangle(bounds.X + 6, bounds.Y + 1, bounds.Width - 12, Math.Max(8, bounds.Height / 3));
        using var shineBrush = new LinearGradientBrush(
            shine,
            Color.FromArgb(85, 110, 170, 215),
            Color.FromArgb(10, 70, 110, 150),
            LinearGradientMode.Vertical);
        using var shinePath = RoundedRect(shine, Math.Max(4, radius - 4));
        graphics.FillPath(shineBrush, shinePath);
        graphics.Restore(state);
    }

    public static void InsetWell(Graphics graphics, Rectangle bounds, int radius)
    {
        FillRound(graphics, bounds, radius, Color.FromArgb(85, 18, 45, 70), Color.FromArgb(115, 10, 30, 50));
        using var highlight = new Pen(Color.FromArgb(95, 90, 150, 215), 1.2f);
        using var path = RoundedRect(Rectangle.Inflate(bounds, -1, -1), Math.Max(2, radius - 1));
        graphics.DrawPath(highlight, path);

        using var innerGlow = new Pen(Color.FromArgb(50, 60, 110, 150), 0.6f);
        using var innerPath = RoundedRect(Rectangle.Inflate(bounds, -2, -2), Math.Max(2, radius - 2));
        graphics.DrawPath(innerGlow, innerPath);
    }

    public static void GlossyButton(Graphics graphics, Rectangle bounds, Color top, Color bottom)
    {
        using (var shadow = RoundedRect(new Rectangle(bounds.X, bounds.Y + 3, bounds.Width, bounds.Height), 9))
        using (var shadowFill = new SolidBrush(Color.FromArgb(45, 12, 28, 45)))
        {
            graphics.FillPath(shadowFill, shadow);
        }

        FillRound(graphics, bounds, 9, top, bottom);
        using var shine = new LinearGradientBrush(
            bounds,
            Color.FromArgb(145, 110, 170, 220),
            Color.FromArgb(15, 70, 110, 150),
            LinearGradientMode.Vertical);
        graphics.FillRectangle(shine, bounds.X + 4, bounds.Y + 1, bounds.Width - 8, bounds.Height / 2);
        using var edge = new Pen(Color.FromArgb(165, 90, 150, 215), 1.3f);
        using var path = RoundedRect(bounds, 9);
        graphics.DrawPath(edge, path);

        using var innerEdge = new Pen(Color.FromArgb(70, 80, 140, 180), 0.7f);
        using var innerPath = RoundedRect(Rectangle.Inflate(bounds, -2, -2), 7);
        graphics.DrawPath(innerEdge, innerPath);
    }

    public static void SectionLabel(Graphics graphics, string text, int x, int y)
    {
        using var font = new Font("Courier New", 7f, FontStyle.Bold);
        using var brush = new SolidBrush(Palette.InkMuted);
        graphics.DrawString(text, font, brush, x, y);
    }

    public static void CenteredText(Graphics graphics, string text, Font font, Brush brush, Rectangle bounds)
    {
        using var format = new StringFormat
        {
            Alignment = StringAlignment.Center,
            LineAlignment = StringAlignment.Center,
            FormatFlags = StringFormatFlags.NoWrap
        };
        graphics.DrawString(text, font, brush, bounds, format);
    }

    public static GraphicsPath RoundedRect(Rectangle bounds, int radius)
    {
        int diameter = Math.Max(2, radius * 2);
        diameter = Math.Min(diameter, Math.Min(bounds.Width, bounds.Height));
        var path = new GraphicsPath();
        path.AddArc(bounds.X, bounds.Y, diameter, diameter, 180, 90);
        path.AddArc(bounds.Right - diameter, bounds.Y, diameter, diameter, 270, 90);
        path.AddArc(bounds.Right - diameter, bounds.Bottom - diameter, diameter, diameter, 0, 90);
        path.AddArc(bounds.X, bounds.Bottom - diameter, diameter, diameter, 90, 90);
        path.CloseFigure();
        return path;
    }

    public static GraphicsPath RoundedTopRect(Rectangle bounds, int radius)
    {
        int diameter = Math.Max(2, radius * 2);
        diameter = Math.Min(diameter, Math.Min(bounds.Width, bounds.Height));
        var path = new GraphicsPath();
        path.AddArc(bounds.X, bounds.Y, diameter, diameter, 180, 90);
        path.AddArc(bounds.Right - diameter, bounds.Y, diameter, diameter, 270, 90);
        path.AddLine(bounds.Right, bounds.Y + radius, bounds.Right, bounds.Bottom);
        path.AddLine(bounds.Right, bounds.Bottom, bounds.X, bounds.Bottom);
        path.AddLine(bounds.X, bounds.Bottom, bounds.X, bounds.Y + radius);
        path.CloseFigure();
        return path;
    }
}

File: src/drawing/Gfx.ts
import { Palette } from './Palette';

export class Gfx {
  static fillRound(
    ctx: CanvasRenderingContext2D,
    bounds: { x: number; y: number; width: number; height: number },
    _radius: number,
    top: string,
    bottom: string
  ): void {
    if (bounds.width <= 0 || bounds.height <= 0) {
      return;
    }
    this.pixelBevel(ctx, bounds, top, bottom, false);
  }

  static glassCard(
    ctx: CanvasRenderingContext2D,
    bounds: { x: number; y: number; width: number; height: number },
    _radius: number
  ): void {
    ctx.fillStyle = Palette.pixelShadow;
    ctx.fillRect(bounds.x + 3, bounds.y + 3, bounds.width, bounds.height);

    const gradient = ctx.createLinearGradient(bounds.x, bounds.y, bounds.x, bounds.y + bounds.height);
    gradient.addColorStop(0, 'rgba(42, 106, 160, 0.94)');
    gradient.addColorStop(1, 'rgba(18, 54, 92, 0.94)');
    ctx.fillStyle = gradient;
    ctx.fillRect(bounds.x, bounds.y, bounds.width, bounds.height);

    ctx.strokeStyle = Palette.pixelNavy;
    ctx.lineWidth = 3;
    ctx.strokeRect(bounds.x + 1.5, bounds.y + 1.5, bounds.width - 3, bounds.height - 3);

    ctx.fillStyle = 'rgba(200, 236, 255, 0.28)';
    ctx.fillRect(bounds.x + 3, bounds.y + 3, bounds.width - 6, 2);
    ctx.fillRect(bounds.x + 3, bounds.y + 3, 2, bounds.height - 6);

    ctx.fillStyle = 'rgba(7, 24, 40, 0.35)';
    ctx.fillRect(bounds.x + bounds.width - 5, bounds.y + 4, 2, bounds.height - 7);
    ctx.fillRect(bounds.x + 4, bounds.y + bounds.height - 5, bounds.width - 8, 2);

    this.scanlines(ctx, bounds, 0.08);
  }

  static insetWell(
    ctx: CanvasRenderingContext2D,
    bounds: { x: number; y: number; width: number; height: number },
    _radius: number
  ): void {
    ctx.fillStyle = '#082038';
    ctx.fillRect(bounds.x, bounds.y, bounds.width, bounds.height);
    ctx.strokeStyle = Palette.pixelNavy;
    ctx.lineWidth = 3;
    ctx.strokeRect(bounds.x + 1.5, bounds.y + 1.5, bounds.width - 3, bounds.height - 3);
    ctx.fillStyle = 'rgba(4, 16, 24, 0.55)';
    ctx.fillRect(bounds.x + 3, bounds.y + 3, bounds.width - 6, 2);
    ctx.fillRect(bounds.x + 3, bounds.y + 3, 2, bounds.height - 6);
    ctx.fillStyle = 'rgba(74, 160, 208, 0.35)';
    ctx.fillRect(bounds.x + 3, bounds.y + bounds.height - 5, bounds.width - 6, 2);
  }

  static glossyButton(
    ctx: CanvasRenderingContext2D,
    bounds: { x: number; y: number; width: number; height: number },
    top: string,
    bottom: string
  ): void {
    this.pixelBevel(ctx, bounds, top, bottom, false);
  }

  static pixelBevel(
    ctx: CanvasRenderingContext2D,
    bounds: { x: number; y: number; width: number; height: number },
    top: string,
    bottom: string,
    selected = false
  ): void {
    ctx.fillStyle = Palette.pixelShadow;
    ctx.fillRect(bounds.x + 3, bounds.y + 3, bounds.width, bounds.height);

    const gradient = ctx.createLinearGradient(bounds.x, bounds.y, bounds.x, bounds.y + bounds.height);
    gradient.addColorStop(0, this.lightenColor(top, 18));
    gradient.addColorStop(1, bottom);
    ctx.fillStyle = gradient;
    ctx.fillRect(bounds.x, bounds.y, bounds.width, bounds.height);

    ctx.strokeStyle = Palette.pixelNavy;
    ctx.lineWidth = 3;
    ctx.strokeRect(bounds.x + 1.5, bounds.y + 1.5, bounds.width - 3, bounds.height - 3);

    ctx.fillStyle = selected ? 'rgba(244, 251, 255, 0.55)' : 'rgba(200, 236, 255, 0.4)';
    ctx.fillRect(bounds.x + 3, bounds.y + 3, bounds.width - 6, 2);
    ctx.fillRect(bounds.x + 3, bounds.y + 3, 2, bounds.height - 6);
  }

  static scanlines(
    ctx: CanvasRenderingContext2D,
    bounds: { x: number; y: number; width: number; height: number },
    alpha = 0.1
  ): void {
    ctx.save();
    ctx.beginPath();
    ctx.rect(bounds.x, bounds.y, bounds.width, bounds.height);
    ctx.clip();
    ctx.fillStyle = `rgba(8, 24, 40, ${alpha})`;
    for (let y = Math.floor(bounds.y); y < bounds.y + bounds.height; y += 4) {
      ctx.fillRect(bounds.x, y, bounds.width, 2);
    }
    ctx.restore();
  }

  static dither(
    ctx: CanvasRenderingContext2D,
    bounds: { x: number; y: number; width: number; height: number },
    color = 'rgba(8, 24, 40, 0.12)'
  ): void {
    ctx.save();
    ctx.beginPath();
    ctx.rect(bounds.x, bounds.y, bounds.width, bounds.height);
    ctx.clip();
    ctx.fillStyle = color;
    for (let y = Math.floor(bounds.y); y < bounds.y + bounds.height; y += 2) {
      for (let x = Math.floor(bounds.x) + (y % 4 === 0 ? 0 : 2); x < bounds.x + bounds.width; x += 4) {
        ctx.fillRect(x, y, 2, 2);
      }
    }
    ctx.restore();
  }

  static sectionLabel(ctx: CanvasRenderingContext2D, text: string, x: number, y: number): void {
    ctx.font = `bold 9px ${Palette.pixelFont}`;
    ctx.fillStyle = Palette.inkMuted;
    ctx.textBaseline = 'top';
    ctx.textAlign = 'left';
    ctx.fillText(text, x, y);
  }

  static centeredText(
    ctx: CanvasRenderingContext2D,
    text: string,
    font: string,
    brush: string,
    bounds: { x: number; y: number; width: number; height: number }
  ): void {
    const previousAlign = ctx.textAlign;
    const previousBaseline = ctx.textBaseline;
    ctx.font = font;
    ctx.fillStyle = brush;
    ctx.textAlign = 'center';
    ctx.textBaseline = 'middle';
    ctx.fillText(text, bounds.x + bounds.width / 2, bounds.y + bounds.height / 2);
    ctx.textAlign = previousAlign;
    ctx.textBaseline = previousBaseline;
  }

  static roundedRect(
    bounds: { x: number; y: number; width: number; height: number },
    radius: number
  ): Path2D {
    const diameter = Math.max(2, radius * 2);
    const clampedDiameter = Math.min(diameter, Math.min(bounds.width, bounds.height));
    const path = new Path2D();
    path.moveTo(bounds.x + clampedDiameter / 2, bounds.y);
    path.arcTo(bounds.x + bounds.width, bounds.y, bounds.x + bounds.width, bounds.y + bounds.height, clampedDiameter / 2);
    path.arcTo(
      bounds.x + bounds.width,
      bounds.y + bounds.height,
      bounds.x,
      bounds.y + bounds.height,
      clampedDiameter / 2
    );
    path.arcTo(bounds.x, bounds.y + bounds.height, bounds.x, bounds.y, clampedDiameter / 2);
    path.arcTo(bounds.x, bounds.y, bounds.x + bounds.width, bounds.y, clampedDiameter / 2);
    path.closePath();
    return path;
  }

  static roundedTopRect(
    bounds: { x: number; y: number; width: number; height: number },
    radius: number
  ): Path2D {
    const diameter = Math.max(2, radius * 2);
    const clampedDiameter = Math.min(diameter, Math.min(bounds.width, bounds.height));
    const path = new Path2D();
    path.moveTo(bounds.x + clampedDiameter / 2, bounds.y);
    path.arcTo(bounds.x + bounds.width, bounds.y, bounds.x + bounds.width, bounds.y + bounds.height, clampedDiameter / 2);
    path.lineTo(bounds.x + bounds.width, bounds.y + bounds.height);
    path.lineTo(bounds.x, bounds.y + bounds.height);
    path.lineTo(bounds.x, bounds.y + clampedDiameter / 2);
    path.arcTo(bounds.x, bounds.y, bounds.x + bounds.width, bounds.y, clampedDiameter / 2);
    path.closePath();
    return path;
  }

  private static lightenColor(color: string, amount: number): string {
    const rgba = this.parseRgba(color);
    return `rgba(${Math.min(255, rgba.r + amount)}, ${Math.min(255, rgba.g + amount)}, ${Math.min(255, rgba.b + amount)}, ${rgba.a})`;
  }

  private static parseRgba(color: string): { r: number; g: number; b: number; a: number } {
    const hex = color.match(/^#([0-9a-f]{6})$/i);
    if (hex) {
      const n = parseInt(hex[1], 16);
      return { r: (n >> 16) & 255, g: (n >> 8) & 255, b: n & 255, a: 1 };
    }
    const match = color.match(/rgba?\((\d+),\s*(\d+),\s*(\d+)(?:,\s*([\d.]+))?\)/);
    if (!match) {
      return { r: 0, g: 0, b: 0, a: 1 };
    }
    return {
      r: parseInt(match[1], 10),
      g: parseInt(match[2], 10),
      b: parseInt(match[3], 10),
      a: match[4] ? parseFloat(match[4]) : 1,
    };
  }
}
